crates/net/eth-wire/src/types/message.rs

```diff
@@ -214,6 +216,39 @@ impl EthMessage {
     }
 }
 
+trait EncodableExt {
+    fn encode_max(&self, size: usize) -> alloy_rlp::Result<Vec<u8>, P2PStreamError>;
+
+    fn encode_truncate(&self, limit: usize) -> alloy_rlp::Result<Vec<u8>>;
+}
+
+impl<T: Encodable> EncodableExt for Vec<T> {
+    fn encode_max(&self, limit: usize) -> alloy_rlp::Result<Vec<u8>, P2PStreamError> {
+        let mut buffer = Vec::new();
+        for item in self {
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                return Err(P2PStreamError::HandshakeError(P2PHandshakeError::DecodeError(
+                    alloy_rlp::Error::Custom("Size limit exceeded"),
+                )));
+            }
+        }
+        Ok(buffer)
+    }
+
+    fn encode_truncate(&self, limit: usize) -> alloy_rlp::Result<Vec<u8>> {
+        let mut buffer = Vec::new();
+        for item in self {
+            item.encode(&mut buffer);
+            if buffer.len() > limit {
+                buffer.truncate(limit);
+                break;
+            }
```

Comment:
this is slightly more complex

the logic should be:
encode all items into buffer if putting an item into buffer leads to: limit, split off the last item and return the Vec
so we need to keep track of the len of the buffer before inserting a new item, then split it off there if it exceeds

Reply:
@emhane  regarding `let current_len = buffer.len();` after the conditional statement you said. Above maybe relevant

Reply:
yes I meant after the line of code, inside the scope opened by the `true` branch of the conditional statement